### ISSUE TITLE
Question: Is it possible to implement GGPO rollback into the emulator?

### ISSUE THREAD

**Issue by KuroRaiden** [NovaSquirrel/Mesen-SX#116]
I love the use of this emulator more over SNES9x. The only reason I'm still using that dinosaur of an emulator is because the build they use on fightcade was retrofitted with rollback netcode (which is pretty rough). I've been a big fan of Messen since I found out about Messen-S in 2019 and even gotten JWong to play some fighting games on it with me for his streams. While we didn't really have any issue a side from random crashes/desyncs, its hard to get other folks to get into it seriously because of the fact that the netplay is more like a remote session to the host PC. I personally use this emulator when I had friends over before I gotten my MiSTerFGPA due to how accurate it is and overall the better quality compared to SNES9X.  I've tried myself to see how to poke around with ARES and even asked in their discord since its similar to MESSEN with its features (not build) and it has N64 support.  I won't lie, I didn't know what the hell I was doing now did I knew how to really implant or translate GGPO's codes into ARES's builds.

TL;DR I really like Messen .

With all that out the way, as the title states, I was curious with how difficult it would be to implement rollback netcode using the [GGPO SDK](https://github.com/pond3r/ggpo)?

**Comment by NovaSquirrel** [comment]
I wouldn't be able to give any advice on the code itself since all I ever did with it was port over the history viewer feature from Mesen, but I'd expect that it's already got most of the work behind rollback in place already just from the other features it has. The rewind feature means it's already storing information on how to go back however many frames, and I think the "run ahead" feature is similar to how rollback works too.

You could leave a suggestion [in this forum thread](https://forums.nesdev.org/viewtopic.php?t=24391) to reach the original author who's working on [the currently maintained version of Mesen](https://github.com/SourMesen/Mesen2), though he's said before the TODO list is huge.

**Comment by KuroRaiden** [comment]
This is HUGE! Thanks for pointing me to the right direction.
